### ISSUE TITLE
Lint to disallow renaming function parameters

### ISSUE THREAD

**Issue by Centri3** [rust-lang/rust-clippy#11443]
### What it does

The name for parameters of a function are entirely arbitrary; however, in many cases the default for that trait is more desirable for consistency's sake.

Should not lint if the pattern is not a single binding or if its name starts with `_` in either the default implementation or the new name.

Should also be allow-by-default.

### Advantage

- Consistency, for example `from_str`'s parameter is `s` which some may wish to always use. Similarly, `From`'s `value`.

### Drawbacks

- It's really really pedantic.

### Example

```rust
struct A(u32);

impl From<A> for String {
    fn from(a: A) -> Self {
        a.0.to_string()
    }
}
```

Could be written as:

```rust
struct A(u32);

impl From<A> for String {
    fn from(value: A) -> Self {
        value.0.to_string()
    }
}
```

**Comment by Centri3** [comment]
By very very pedantic, I didn't mean too pedantic for `pedantic`; I mean it may be too pedantic even for `restriction` ^^

We could allow library authors to opt into this through an attribute, and just annotate those traits with it.

**Comment by J-ZhengLi** [comment]
@rustbot claim

I'll give it a try 😅
